dot/parachain/grid/grid_topology.go

```diff
@@ -0,0 +1,352 @@
+package grid
+
+import (
+	"fmt"
+	"math"
+
+	parachaintypes "github.com/ChainSafe/gossamer/dot/parachain/types"
+	"github.com/ChainSafe/gossamer/dot/types"
+
+	"github.com/libp2p/go-libp2p/core/peer"
+)
+
+// TopologyPeerInfo is an information about the peer in the gossip topology
+type TopologyPeerInfo struct {
+	// Peers is a list pf peers this peer knows about
+	Peers peer.IDSlice
+	/// ValidatorIndex is the index of the validator in the discovery keys of the corresponding
+	/// `SessionInfo`. This can extend _beyond_ the set of active parachain validators.
+	ValidatorIndex parachaintypes.ValidatorIndex
+	/// DiscoveryID is the authority discovery public key of the validator in the corresponding
+	/// `SessionInfo`.
+	DiscoveryID types.AuthorityID
+}
+
+// SessionGridTopology is topology representation for session
+type SessionGridTopology struct {
+	// Peers List of all peers ids in session. Represented as "hashset" for fast lookup.
+	Peers map[peer.ID]struct{}
+	/// canonicalShuffling is the canonical shuffling of validators for the session.
+	CanonicalShuffling []TopologyPeerInfo
+	/// shuffledIndices is an array mapping validator indices to their indices in the
+	/// shuffling itself. This has the same size as the number of validators
+	/// in the session.
+	ShuffledIndices []uint
+}
+
+// NewSessionGridTopology creates a new SessionGridTopology
+// Peers populated with peer ids from canonicalShuffling argument.
+// canonicalShuffling set of authorities for this session. within one session it can have only one key.
+func NewSessionGridTopology(shuffledIndices []uint, canonicalShuffling []TopologyPeerInfo) *SessionGridTopology {
+	peers := make(map[peer.ID]struct{})
+	for _, peerInfo := range canonicalShuffling {
+		for _, p := range peerInfo.Peers {
+			peers[p] = struct{}{}
+		}
+	}
+	return &SessionGridTopology{
+		Peers:              peers,
+		CanonicalShuffling: canonicalShuffling,
+		ShuffledIndices:    shuffledIndices,
+	}
+}
+
+// UpdateAuthoritiesIDs Updates the known peer ids in SessionGridTopology for the passed authorities ids.
+// Between sessions validators can update their authorityID because of key rotation, peerID might be changed as well.
+// hence there could be multiple AuthorityID associated with a peerID.
+// Updates Peers hashset with new peer id if the peer is in the grid topology.
+// Returns true if the peer is in the grid topology.
+func (gt *SessionGridTopology) UpdateAuthoritiesIDs(peer peer.ID, discoveryIDs map[types.AuthorityID]struct{}) bool {
+	updated := false
+	if _, ok := gt.Peers[peer]; !ok {
+		for i := range gt.CanonicalShuffling {
+			p := &gt.CanonicalShuffling[i]
+			if _, ok := discoveryIDs[p.DiscoveryID]; ok {
+				gt.Peers[peer] = struct{}{}
+				p.Peers = append(p.Peers, peer)
+				updated = true
+			}
+		}
+	}
+	return updated
+}
```

Comment:
Will this return true even if there is only one authority ID in discoveryIDs? Is that the intended behavior?

Reply:
There is actually as I understood in most cases will be only one authorityID. But yes it should be true if even one update occure